software/apps/axpy/main.c

```diff
@@ -0,0 +1,209 @@
+  return 0;
+}
+
+// Serial calculation by core 0
+void calc_axpy_serial(int32_t *matrix_X, int32_t *matrix_Y, int32_t alpha,
+                      uint32_t elements, uint32_t core_id) {
+  if (core_id == 0) {
+    AXPY(elements, alpha, &matrix_X[0], &matrix_Y[0]);
+  }
+}
+
+// Serial calculation with 4 time unloop by core 0
+void calc_axpy_serial_unloop(int32_t *matrix_X, int32_t *matrix_Y,
+                             int32_t alpha, uint32_t elements,
+                             uint32_t core_id) {
+  if (core_id == 0) {
+    AXPY_unloop(elements, alpha, &matrix_X[0], &matrix_Y[0]);
+  }
+}
+
+// Parallel calculation
+void calc_axpy(int32_t *matrix_X, int32_t *matrix_Y, int32_t alpha,
+               uint32_t elements, uint32_t core_id, uint32_t num_cores) {
+  // Support the elements number is not the devided by the core numbers;
+  // The corresponding core ID will take the partial elements.
+  uint32_t split = elements / num_cores;
+  uint32_t partial = elements % num_cores;
+  if (core_id < partial) {
+    uint32_t const c_start = core_id * (split + 1);
+    uint32_t const j = split + 1;
+    AXPY(j, alpha, &matrix_X[c_start], &matrix_Y[c_start]);
+  } else {
+    uint32_t const c_start = core_id * split + partial;
+    AXPY(split, alpha, &matrix_X[c_start], &matrix_Y[c_start]);
+  }
+}
+
+// Parallel calculation with 4 times unloop
+void calc_axpy_unloop(int32_t *matrix_X, int32_t *matrix_Y, int32_t alpha,
+                      uint32_t elements, uint32_t core_id, uint32_t num_cores) {
+  // Support the elements number is not the devided by the core numbers;
+  // The corresponding core ID will take the partial elements.
+  uint32_t split = elements / num_cores;
+  uint32_t partial = elements % num_cores;
+  if (core_id < partial) {
+    uint32_t const c_start = core_id * (split + 1);
+    uint32_t const j = split + 1;
+    AXPY_unloop(j, alpha, &matrix_X[c_start], &matrix_Y[c_start]);
+  } else {
+    uint32_t const c_start = core_id * split + partial;
+    AXPY_unloop(split, alpha, &matrix_X[c_start], &matrix_Y[c_start]);
+  }
+}
+
+// Parallel calculation with 4 times unloop align with banks
+void calc_axpy_unloop_x4_localbank(int32_t *matrix_X, int32_t *matrix_Y,
+                                   int32_t alpha, uint32_t elements,
+                                   uint32_t core_id, uint32_t num_cores) {
+  uint32_t const bank_num = num_cores * 4;
+  // Do the calculation that redundant elements cannot be unloop;
+  // Use core0 is less overhead than found the local
+  uint32_t partial = elements % 4;
+  if (core_id == 0) {
+    if (partial != 0) {
+      uint32_t c_start = elements - partial + 1;
+      AXPY(partial, alpha, &matrix_X[c_start], &matrix_Y[c_start]);
+    }
+  }
+  // Do unloop 4 times
+  uint32_t const total_unloop = elements - partial;
+  uint32_t const c_start = core_id * 4;
+  for (uint32_t c = c_start; c <= total_unloop - 4; c += bank_num) {
+    AXPY_unloop_x4(alpha, &matrix_X[c], &matrix_Y[c]);
+  }
+}
```

Comment:
Please move those functions into the `axpy.h` file.

Reply:
Sorry for forgetting this last time.